common/src/parse.cpp

```diff
@@ -79,60 +81,90 @@ pcl::console::parse_argument (int argc, const char * const * argv, const char *
 int
 pcl::console::parse_argument (int argc, const char * const * argv, const char * str, bool &val)
 {
+  long int dummy;
+  const auto ret = parse_argument (argc, argv, str, dummy);
+  if (ret != -1)
+  {
+    val = static_cast<bool> (dummy);
+  }
+  return ret;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+namespace pcl
+{
+namespace console
+{
+template <class T, class V = T(*)(const char*, const char**)> int
+parse_generic (V convert_func, int argc, const char* const* argv, const char* str, T& val)
+{
+  char *endptr = nullptr;
   int index = find_argument (argc, argv, str) + 1;
+  errno = 0;
 
   if (index > 0 && index < argc )
-    val = atoi (argv[index]) == 1;
+  {
+    val = convert_func (argv[index], &endptr);  // similar to strtol, strtod, strtof
+  }
+  // handle out-of-range, junk at the end and no conversion
+  if (errno == ERANGE || *endptr != '\0' || str == endptr)
+  {
+    return -1;
+  }
 
   return (index - 1);
 }
 
+template <class Integral> int
+parse_longint (int argc, const char * const * argv, const char * str, Integral &val)
```

Comment:
The name of this function is marginally misleading. It is especially funny because you use it to parse all integers but the ones of `long` type. Suggestion to rename it to `parse_integral`?

Reply:
I want to tell it can be used to parse integrals upto size `long int`. `parse_integral` doesn't convey that (since `unsigned long` and `long long` are not addressed). Maybe SFINAE?

I wanted to use `parse_max_longint` but due to use of numeric limits, thought it would be confusing. Let me look at the codegen and if it's same at -O2, I can then route the `long int` function through here too.

Reply:
PTAL. I have ensured `parse_longint` is used for `long int` as well as only for valid integral types (not for short int, should add that though)